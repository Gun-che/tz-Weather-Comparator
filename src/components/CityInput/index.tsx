import React, { SyntheticEvent } from 'react'
import s from './index.module.scss'

interface ICityInputProps {
  index: number;
  onChange(e: React.ChangeEvent<HTMLInputElement & { dataset: { index: number } }>): void;
  item: [string, number];
  onClose(index: number): (e: SyntheticEvent) => void;
}

export const CityInput: React.FC<ICityInputProps> = ({
  index,
  item,
  onChange,
  onClose
}) => {

  return (
    <div className={"input-field " + s.wrap}>
      <button
        className="btn-floating btn-small waves-effect waves-light blue lighten-3"
        onClick={onClose(index)}>
        <i role="button" className="material-icons">close</i>
      </button>
      <i className="material-icons prefix">location_on</i>
      <input
        className="validate"
        data-index={index}
        type="text"
        onChange={onChange}
        id={'city' + (index + 1)}
        name={'city' + (index + 1)}
        value={item[0]}
      />
      <label htmlFor={'city' + (index + 1)}>{index + 1 + ' город'}</label>
    </div>
  )
}
