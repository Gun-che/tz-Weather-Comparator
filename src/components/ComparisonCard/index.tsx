import React from 'react'
import { IWeatherData } from '../../types/state'
import s from './index.module.scss'
import img from './wp.jpg'

export const ComparisonField: React.FC<{ data: IWeatherData[] }> = ({ data }) => {

  const sortByTemp = (): [string, number, number] => {
    let temp = -50
    let feelsLike = 0
    let city = ''
    data.forEach(i => {
      if (i.main.feels_like > temp) {
        temp = i.main.temp;
        city = i.name;
        feelsLike = i.main.feels_like
      }
    })
    return [city, temp, feelsLike]
  }

  return (
    <div className={s.wrap}>
      <div className="card">
        <div className="card-image">
          <img src={img} alt="sky" />
          <span className="card-title">{sortByTemp()[0]} cамый тёплый город, из выбранных вами!</span>
        </div>
        <div className="card-content">
          <p>Температура здесь сегодня {Math.round(sortByTemp()[1]) + `°C!`}</p>
          <p>По ощущениям {Math.round(sortByTemp()[2]) + `°C`}</p>
        </div>

      </div>
      <h2>{}</h2>
    </div>
  )
}
