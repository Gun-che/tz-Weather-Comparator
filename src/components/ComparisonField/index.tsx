import React from 'react'
import { IWeatherData } from '../../types/state'

export const ComparisonField: React.FC<{ data: IWeatherData[] }> = ({ data }) => {
  return (
    <div>

      {data.length ? <h2>{data[0].main.temp} ? {data[1].main.temp}</h2> : null}
    </div>
  )
}
