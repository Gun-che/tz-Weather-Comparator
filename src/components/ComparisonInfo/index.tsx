import React from 'react'
import { IWeatherData } from '../../types/state'

interface ISortResult {
  warmCity: string;
  coldCity: string;
  windyCity: string;
  windlessCity: string;
  wetCity: string;
  dryCity: string;
}

export const ComparisonInfo: React.FC<{ data: IWeatherData[] }> = ({ data }) => {

  const sort = (): ISortResult => {
    let warm = -100
    let cold = 100
    let windy = 0
    let windless = 100
    let wet = 0
    let dry = 100
    let warmCity = ''
    let coldCity = ''
    let windyCity = ''
    let windlessCity = ''
    let wetCity = ''
    let dryCity = ''

    data.forEach(i => {

      if (i.main.feels_like > warm) {
        warmCity = i.name;
        warm = i.main.feels_like
      }

      if (i.main.feels_like < cold) {
        coldCity = i.name;
        cold = i.main.feels_like
      }

      if (i.wind.speed > windy) {
        windyCity = i.name;
        windy = i.main.feels_like
      }

      if (i.wind.speed < windless) {
        windlessCity = i.name;
        windless = i.main.feels_like
      }

      if (i.main.humidity > wet) {
        wetCity = i.name;
        wet = i.main.feels_like
      }

      if (i.main.humidity < dry) {
        dryCity = i.name;
        dry = i.main.feels_like
      }
    })

    return {
      warmCity,
      coldCity,
      windyCity,
      windlessCity,
      wetCity,
      dryCity,
    }
  }

  const {
    warmCity,
    coldCity,
    windyCity,
    windlessCity,
    wetCity,
    dryCity, }: ISortResult = sort()

  return (
    <div>
      <div className="card blue-grey darken-1">
        <div className="card-content white-text">
          <span className="card-title">Card Title</span>
          <p>Теплее прочего: {warmCity}</p>
          <p>Холоднее прочего: {coldCity}</p>
          <p>Ветренее прочего: {windyCity}</p>
          <p>Наиенее ветренно: {windlessCity}</p>
          <p>Самая высокая влажность: {wetCity}</p>
          <p>Суше прочего: {dryCity}</p>
        </div>
      </div>
    </div>
  )
}
