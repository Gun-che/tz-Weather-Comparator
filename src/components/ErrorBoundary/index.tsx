import React from 'react'
import s from './index.module.scss'

interface IState {
  hasError: boolean;
  msg: string;
  err: number;
}

export class ErrorBoundary extends React.Component<{}, IState> {
  constructor(props: {}) {
    super(props);
    this.state = {
      hasError: false,
      msg: '',
      err: 0,
    }
  }

  static getDerivedStateFromError(error: Error) {
    return {
      hasError: true,
      msg: error.message,
      err: error.name,
    }
  }

  render() {
    const err = this.state.err;
    const msg = this.state.msg;
    const hasError = this.state.err;

    if (hasError) {
      return (
        <>
          <h1>Что-то пошлло не так, пожалуйста обновите страницу!</h1>
          <div className={s.wrap + ' container'}>
            {err && err}
            {msg && <h3>{msg}</h3>}
          </div>
        </>
      )

    } else {
      return this.props.children
    }
  }
}
