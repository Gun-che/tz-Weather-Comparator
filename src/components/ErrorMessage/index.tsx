import React from 'react'
import s from './index.module.scss'

export const ErrorMessage: React.FC<{ message: string }> = (
  { message }
) => {

  return (<>
    {message ? (<div className={s.wrap + ' containerCust'}><h2>{message}</h2></div>) : (<div className={s.emptyWrap}></div>)
    }
  </>)
}
