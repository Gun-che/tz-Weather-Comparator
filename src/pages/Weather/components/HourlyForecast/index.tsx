import React from 'react'
import { IHourlyWeatherData } from '../../../../types/apiResponse'
import { timeFormater } from '../../../../helpers/dateFormater'
import { withACapitalLetter } from '../../../../helpers/withACapitalLetter'
import { weatherMsg } from '../../../../helpers/messages'
import { baseIconUrl } from '../../../../helpers/api'
import s from './index.module.scss'

export const HourlyForecast: React.FC<{ data: IHourlyWeatherData[], name: string }> = ({ data, name }) => {

  return (
    <section className={s.pageWrap + ' animEl'}>
      <h2>Прогноз на неделю в городе {name}</h2>
      <div className={s.wrap}>

        {data.map(({
          weather: [{ id, icon, description }],
          temp,
          wind_speed,
          dt
        }) => {

          const anotation = withACapitalLetter(weatherMsg[id]['ru'])
          const date = timeFormater(dt)

          return (
            <div key={dt} className={s['h' + date.hour] + ' ' + s.sec}>
              <div>
                <p className={s.time}>{date.time}</p>
                <p className={s.date}>{date.date}</p>
              </div>
              <div className={s.img}>
                <img src={`${baseIconUrl}${icon}.png`} alt={description} title={anotation} />
              </div>
              <div className={s.temp}>
                <div className=""><i className="material-icons">flare</i>{Math.round(temp)}°C</div>
                <div><i className="material-icons">navigate_next</i>{wind_speed}м/с</div>
              </div>
            </div>
          )
        })}
      </div>
    </section >
  )
}
