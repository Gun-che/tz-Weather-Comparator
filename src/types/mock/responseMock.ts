import { IDailyWeatherData, IHourlyWeatherData, IWeatherData, IAllWeatherData } from "../apiResponse";

export const weatherDataMock: IWeatherData = {
  name: '1',
  visibility: 1,
  weather: [{
    id: 1,
    icon: '1',
    description: '1',
  }],
  wind: { speed: 1 },
  coord: { lon: 1, lat: 1 },
  main: {
    feels_like: 1,
    humidity: 1,
    pressure: 1,
    temp: 1,
  },
}

export const hourlyWeatherDataMock: IHourlyWeatherData = {
  clouds: 1,
  feels_like: 1,
  humidity: 1,
  pressure: 1,
  temp: 1,
  weather: [{
    id: 1,
    icon: '1',
    description: '1',
  }],
  wind_speed: 1,
  dt: 1,
}

export const dailyWeatherDataMock: IDailyWeatherData = {
  clouds: 10,
  weather: [{
    id: 1,
    icon: '1',
    description: '1',
  }],
  pressure: 10,
  feels_like: { day: 1, night: 1, eve: 1, morn: 1 },
  humidity: 1,
  sunrise: 1,
  sunset: 1,
  temp: { day: 1, night: 1, eve: 1, morn: 1 },
  wind_speed: 1,
  dt: 1,
}

export const allWeatherDataMock: IAllWeatherData = {
  hourly: [hourlyWeatherDataMock],
  daily: [dailyWeatherDataMock],
}