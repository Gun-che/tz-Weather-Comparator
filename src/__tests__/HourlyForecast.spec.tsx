import React from 'react'
import { shallow } from 'enzyme'
import { HourlyForecast } from '../pages/Weather/components/HourlyForecast'
import { hourlyWeatherDataMock } from '../types/mock/responseMock'

describe('Hourly Forecast component', () => {
  const props = {
    data: [hourlyWeatherDataMock, hourlyWeatherDataMock],
    name: 'moscow'
  }
  describe('hourlyForecast component render', () => {
    const hourlyForecast = shallow(<HourlyForecast {...props} />)

    it('renders properly', () => {
      expect(hourlyForecast).toMatchSnapshot()
    })

    it('render forecast item', () => {
      expect(hourlyForecast.find('div.sec')).toHaveLength(props.data.length)
    })

    it('render <h2>', () => {
      expect(hourlyForecast.find('h2')).toHaveLength(1)
    })

    it('renders inner <h2>', () => {
      expect(hourlyForecast
        .find('h2')
        .first()
        .text()).toEqual(`Почасовой прогноз на 2 дня в городе ${props.name}`)
    })
  })
})