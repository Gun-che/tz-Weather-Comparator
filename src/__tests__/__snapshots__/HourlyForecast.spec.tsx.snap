// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hourly Forecast component hourlyForecast component render renders properly 1`] = `
<section
  className="pageWrap animEl"
>
  <h2>
    Почасовой прогноз на 2 дня в городе 
    moscow
  </h2>
  <div
    className="wrap"
  >
    <div
      className="h3 sec"
      key="1"
    >
      <div>
        <p
          className="time"
        >
          03:00
        </p>
        <p
          className="date"
        >
          1 ├F3: M01┤
        </p>
      </div>
      <div
        className="img"
      >
        <img
          alt="1"
          src="http://openweathermap.org/img/wn/1.png"
          title="Безоблачно"
        />
      </div>
      <div
        className="temp"
      >
        <div
          className=""
        >
          <i
            className="material-icons"
          >
            flare
          </i>
          1
          °C
        </div>
        <div>
          <i
            className="material-icons"
          >
            navigate_next
          </i>
          1
          м/с
        </div>
      </div>
    </div>
    <div
      className="h3 sec"
      key="1"
    >
      <div>
        <p
          className="time"
        >
          03:00
        </p>
        <p
          className="date"
        >
          1 ├F3: M01┤
        </p>
      </div>
      <div
        className="img"
      >
        <img
          alt="1"
          src="http://openweathermap.org/img/wn/1.png"
          title="Безоблачно"
        />
      </div>
      <div
        className="temp"
      >
        <div
          className=""
        >
          <i
            className="material-icons"
          >
            flare
          </i>
          1
          °C
        </div>
        <div>
          <i
            className="material-icons"
          >
            navigate_next
          </i>
          1
          м/с
        </div>
      </div>
    </div>
  </div>
</section>
`;
